### PR TITLE
Improve statistical indicators

### PR DESCRIPTION
**Pull request by bookjiang**
### Motivation
"Improved message write statistics for topics" is  a program of "Summer 2021 of Open Source Promotion Plan" :  Restores the number of writes to the topic when the broker restarts or the topic changes.

### Modifications
- *I add the _**getLastEntry()**_ method  in **_ManagedLedgerImpl.java_** to  get the last written data for a topic.*
- *I add some code in **_BrokerService.java_** to restore the number of writes when broker load the topic.*
- *I add four test function in **_BrokerServiceTest.java_** to test the function I add*

### Verifying this change

- [ ] Make sure that the change passes the CI checks.

*(Please pick either of the following options)*

This change added tests and can be verified as follows:

- *Restore the number of writes **_without batch_**  when **_broker restarts_**.*
- *Restore the number of writes **_without batch_**  when **_broker restarts_** and sending messages  **_asynchronously_***
- *Restore the number of writes **_with batch_**  when **_broker restarts_** and sending messages  **_asynchronously_***
- *Restore the number of writes **_with batch_**  when **_topic is unloaded_**  and sending messages  **_asynchronously_***

### Does this pull request potentially affect one of the following parts:

*If `yes` was chosen, please highlight the changes*

  - Dependencies (does it add or upgrade a dependency): (no)
  - The public API: ( no)
  - The schema: (no )
  - The default values of configurations: (no)
  - The wire protocol: (no)
  - The rest endpoints: (no)
  - The admin cli options: (no)
  - Anything that affects deployment: (no)

### Documentation

Check the box below and label this PR (if you have committer privilege).

Need to update docs? 

- [ ] doc-required 
  
  (If you need help on updating docs, create a doc issue)
  
- [x] no-need-doc 
  
  (Please explain why)
  
- [ ] doc 
  
  (If this PR contains doc changes)

### CODE REVIEW

File: pulsar-broker/src/main/java/org/apache/pulsar/broker/service/BrokerService.java [hunk 1]
```diff
@@ -1248,6 +1252,21 @@ public void openLedgerComplete(ManagedLedger ledger, Object ctx) {
                                             topics.remove(topic, topicFuture);
                                         });
                                     } else {
+                                        ManagedLedgerImpl managedLedger =
+                                                (ManagedLedgerImpl) persistentTopic.getManagedLedger();
+                                        CompletableFuture<LedgerEntry> entryFuture = managedLedger.getLastEntry();
+                                        entryFuture.whenCompleteAsync((ledgerEntry, e)->{
+                                            if (e != null) {
+                                                log.error("fetch lastEntry error---{}", e.getMessage());
```
Comment (gaoran10) [inline pulsar-broker/src/main/java/org/apache/pulsar/broker/service/BrokerService.java:1260]:
```suggestion
                                                log.error("Failed to get last entry {}.", managedLedger.getLastConfirmedEntry(), e);
```

Comment (gaoran10) [reply pulsar-broker/src/main/java/org/apache/pulsar/broker/service/BrokerService.java:1260]: If message count recovery failed, the topic creation should fail.

**Review by gaoran10 — COMMENTED** [review]
Thanks for your contribution.

Left some comments.

1. It seems that there are some changed files that don't belong to this PR, please rebase with branch `apache/master`.
2. Please check the class `AppendIndexMetadataInterceptor`, it is used to record the total message count for each topic and it already has a method `recoveryIndexGenerator`, we could get the message total count from this class.

File: managed-ledger/src/main/java/org/apache/bookkeeper/mledger/impl/ManagedLedgerImpl.java [hunk 2]
```diff
@@ -3964,5 +3966,19 @@ private void updateLastLedgerCreatedTimeAndScheduleRolloverTask() {
                     safeRun(this::rollCurrentLedgerIfFull), this.maximumRolloverTimeMs, TimeUnit.MILLISECONDS);
         }
     }
+    public CompletableFuture<LedgerEntry> getLastEntry() {
+        if (lastConfirmedEntry.getEntryId() == -1){
+            return new CompletableFuture<>();
```
[the remark this answers is not in the document]
Comment (bookjiang) [reply managed-ledger/src/main/java/org/apache/bookkeeper/mledger/impl/ManagedLedgerImpl.java:3971]: Thank you, teacher, it has been corrected.

**Comment by tisonkun** [comment]
Closed as stale and conflict. Please rebase and resubmit the patch if it's still relevant.

**Pull request closed without merging**